source/Halibut/Transport/Protocol/TemporaryFileDataStreamReceiver.cs

```diff
@@ -0,0 +1,36 @@
+﻿using System;
+using System.IO;
+
+namespace Halibut.Transport.Protocol
+{
+    public class TemporaryFileDataStreamReceiver : IDataStreamReceiver
+    {
+        readonly Action<Stream> writer;
+
+        public TemporaryFileDataStreamReceiver(Action<Stream> writer)
+        {
+            this.writer = writer;
+        }
+
+        public void SaveTo(string filePath)
+        {
+            using (var file = new FileStream(filePath, FileMode.Create))
+            {
+                writer(file);
+            }
+        }
+
+        public void Read(Action<Stream> reader)
+        {
+            var path = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
+            using (var fileStream = new FileStream(path, FileMode.Create, FileAccess.ReadWrite, FileShare.None, 65536))
+            {
+                writer(fileStream);
+                fileStream.Seek(0, SeekOrigin.Begin);
+                reader(fileStream);
+            }
+            if (File.Exists(path))
```

Comment:
maybe wrap this in a try statement with the delete in a finally. just to make sure if it fails  to transfer we still try and delete.